src/log_internal.h

```diff
@@ -123,54 +123,77 @@ class FileLogOutput : public ICombinedLogOutput {
 	std::ofstream m_stream;
 };
 
-class LogOutputBuffer : public ICombinedLogOutput {
+struct LogEntry {
+	LogLevel level;
+	std::string timestamp;
+	std::string thread_name;
+	std::string text;
+	// "timestamp: level[thread_name]: text"
+	std::string combined;
+
+	LogEntry(LogLevel level, std::string timestamp, std::string thread_name,
+			std::string text, std::string combined) :
+		level(level),
+		timestamp(std::move(timestamp)),
+		thread_name(std::move(thread_name)),
+		text(std::move(text)),
```

Comment:
i'm not sure a std::move on a constructor with copy has any effect

Reply:
The parameter types are just `std::string`. They should be `std::string &&` if you don't want to copy them by pure virtue of calling the constructor.

If you do want to copy them once (looks like that's the case since they are passed on from `log`, which receives const refs, as makes sense because you wouldn't want strings you sent to logging to be consumed), the parameters should be `const std::string &` to be consistent in terms of style.

(Since this is a struct and you want to initialize all fields, why define a constructor at all though? You could just use [aggregate initialization](https://en.cppreference.com/w/cpp/language/aggregate_initialization).)

Reply:
@appgurueu 

If method `foo` is passed a string, and needs to retain a copy of it:

```
void foo(string arg);                // can avoid alloc & copy if the caller gives up theirs (using std::move)
void foo(const string& arg);  // no way to avoid alloc & copy
void foo(string_view arg);      // no way to avoid alloc & copy
```

So I generally use (1) for methods that need to retain a copy of "arg", and (2) or (3) when it doesn't.

In this case, a copy does need to happen, but it happens in log() and logRaw() when invoking the constructor. I'll switch to aggregate initializers here to minimize confusion.

Reply:
We always use `void foo(const string& arg)` when a copy is expected to happen and `void foo(string&& arg)` when it isn't.
I haven't seen the `void foo(string arg)` style anywhere else and I'm not sure if it actually works like that.

Reply:
@sfan5

Here's a snippet using a custom allocator to track allocations:

https://gist.github.com/paradust7/a1fbaaaac049964c49efe22d1fea785f

It outputs:

```
Calling: foo_unified(str1)
Allocating 1048577 bytes
done

Calling: foo_unified(std::move(str2))
done

Calling: foo_split(str3)
Allocating 1048577 bytes
done

Calling: foo_split(std::move(str4))
done
```

i.e. `foo_unified` achieves the same performance as the two separate `foo_split` methods.

There is a small technical difference, in that the internal variables (ptr and length) of the string are swapped twice by `foo_unified`, but only once by `foo_split(&&)`, but this difference is tiny. And I suspect the compiler will be able to optimize it out anyway.

Reply:
In any case, I have removed the code that did this, so it is not longer a blocker/issue for this PR. I just wanted to demonstrate that it does actually do that.